ion-schema/src/constraint.rs

```diff
@@ -2295,3 +2310,73 @@ impl ConstraintValidator for TimestampOffsetConstraint {
         Ok(())
     }
 }
+
+/// Implements Ion Schema's `ieee754_float` constraint
+/// [ieee754_float]: https://amazon-ion.github.io/ion-schema/docs/isl-2-0/spec#ieee754_float
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub struct Ieee754FloatConstraint {
+    interchange_format: Ieee754InterchangeFormat,
+}
+
+impl Ieee754FloatConstraint {
+    pub fn new(interchange_format: Ieee754InterchangeFormat) -> Self {
+        Self { interchange_format }
+    }
+
+    pub fn interchange_format(&self) -> Ieee754InterchangeFormat {
+        self.interchange_format
+    }
+}
+
+impl ConstraintValidator for Ieee754FloatConstraint {
+    fn validate(
+        &self,
+        value: &IonSchemaElement,
+        type_store: &TypeStore,
+        ion_path: &mut IonPath,
+    ) -> ValidationResult {
+        // get ieee interchange format value
+        let float_value = value
+            .expect_element_of_type(&[IonType::Float], "ieee754_float", ion_path)?
+            .as_float()
+            .unwrap();
+
+        if !float_value.is_finite() {
+            return Ok(());
+        }
+
+        match self.interchange_format {
+            Ieee754InterchangeFormat::Binary16 => {
+                let f16_value = half::f16::from_f64(float_value);
+                if f16_value.to_f64() != float_value {
+                    return Err(Violation::new(
+                        "ieee754_float",
+                        ViolationCode::InvalidIeee754Float,
+                        format!("value cannot be losslessly represented by the IEEE-754 {} interchange format.", self.interchange_format),
+                        ion_path,
+                    ));
+                }
+                Ok(())
+            }
+            Ieee754InterchangeFormat::Binary32 => {
+                let f64_value = float_value.to_f32().and_then(|f32_value| f32_value.to_f64()).ok_or(Violation::new(
+                    "ieee754_float",
+                    ViolationCode::InvalidIeee754Float,
+                    format!("value cannot be losslessly represented by the IEEE-754 {} interchange format.", self.interchange_format),
+                    ion_path,
+                ))?;
+
+                if f64_value != float_value {
+                    return Err(Violation::new(
+                        "ieee754_float",
+                        ViolationCode::InvalidIeee754Float,
+                        format!("value cannot be losslessly represented by the IEEE-754 {} interchange format.", self.interchange_format),
+                        ion_path,
+                    ));
+                }
+                Ok(())
+            }
+            Ieee754InterchangeFormat::Binary64 => Ok(()),
+        }
```

Comment:
You can simplify and get rid of repeated code by moving the violations out of the `match` expression. (I haven't checked this example to make sure it's syntactically correct, but this is the general idea.)

```suggestion
        let is_valid = match self.interchange_format {
            Ieee754InterchangeFormat::Binary16 => {
                half::f16::from_f64(float_value).to_f64() == float_value
            }
            Ieee754InterchangeFormat::Binary32 => float_value.to_f32()
                .and_then(|f32_value| f32_value.to_f64() == float_value)
                .ok_or(false),
            Ieee754InterchangeFormat::Binary64 => true,
        }
        if is_valid {
            Ok(())
        } else {
            Err(Violation::new(
                "ieee754_float",
                ViolationCode::InvalidIeee754Float,
                format!("value cannot be losslessly represented by the IEEE-754 {} interchange format.", self.interchange_format),
                ion_path,
            ));
        }
```